File: Week4/closestNumbers/ClosestNumbers.java
package closestNumbers;

import java.util.*;

public class ClosestNumbers {
    public static void main(String[] args){
        List<Integer> numbers = new ArrayList<>();
        numbers.addAll(Arrays.asList(  -20, -3916237, -357920, -3620601, 7374819, -7330761, 30, 6246457, -6461594,
        266854));
        System.out.println(closestNumbers(numbers));
    }
    public static List<Integer> closestNumbers(List<Integer> arr){
        int smallestDiff = Integer.MAX_VALUE;
        List<Integer> pairs = new ArrayList<>();
        List<Integer[]> diffAndPairs = new ArrayList<>();
                
        Collections.sort(arr);
        
        for (int i = 0; i < arr.size() - 1; i++) {
            int difference = Math.abs(arr.get(i + 1) - arr.get(i));
            if (smallestDiff > difference) {
                smallestDiff = difference;
                Integer[] pair = { difference, arr.get(i), arr.get(i + 1) };
                diffAndPairs.add(pair);
            }
        }
        
        for (Integer[] pair : diffAndPairs) {
            if (pair[0].equals(smallestDiff)) {
                pairs.add(pair[1]);
                pairs.add(pair[2]);
            }
        }
        return pairs;
}
}

File: Week4/closestNumbers/closestNumbers.ts
function closestNumbers2(arr:number[]):number[]{
    
    let smalletDif:number|undefined;
    let pairs:number[]=[];
    let diffAndPairs:number[][]=[];
    let sortedArr = arr.sort((a,b)=>a-b)

    for(let i = 0; i < sortedArr.length; i++){
        let difference:number = Math.abs(sortedArr[i+1]-sortedArr[i]);
        if( smalletDif === undefined || smalletDif > difference){
            smalletDif = difference;
            diffAndPairs.push([difference,sortedArr[i],sortedArr[i+1]]);  
        }
    }
    for(let k =0; k < diffAndPairs.length; k++){
        diffAndPairs[k][0] === smalletDif &&
        pairs.push(diffAndPairs[k][1],diffAndPairs[k][2])
    }

    return pairs
}

let numbers2:number[]=[  -20, -3916237, -357920, -3620601, 7374819, -7330761, 30, 6246457, -6461594,
    266854]

console.log(closestNumbers2(numbers2));